doc/sphinx/proof-engine/ltac.rst

```diff
@@ -635,6 +651,31 @@ In some cases backtracking may be too expensive.
 
             Tactic Notation "foo" tactic_list(tacs) := first tacs.
 
+Solving
+~~~~~~~
+
+Selects and applies the first tactic that solves each goal (i.e. leaves no subgoal)
+in a series of alternative tactics:
+
+.. tacn:: solve [ {*| @ltac_expr__i } ]
+   :name: solve
+
+   The :n:`@ltac_expr__i` are evaluated to :n:`v__i` and :n:`v__i` must be
+   tactic values, for i = 1, ..., n. If n > 1,
+   :tacn:`solve` :n:`[@ltac_expr__1 | ... | @ltac_expr__n]` applies :n:`v__1` to
+   each goal independently and stops if it succeeds; otherwise it tries to
+   apply :n:`v__2` and so on. The message :exn:`No applicable tactic` indicates
+   there is no solving tactic.
```

Comment:
Again, evaluation and application are interleaved.  If one of the branches succeeds in solving the goal, the subsequent branches will not even be evaluated.  Additionally, if a branch fails, or fails to solve the goal, not just the effects of application but also the effects of evaluation are rolled back.

Reply:
I have this now:

>   For each current subgoal: evaluates and applies each :n:`@ltac_expr` in order
   until one is found that solves the subgoal.  The :n:`@ltac_expr`\s must evaluate
   to tactic values.
   
>   If any of the subgoals are not solved, then the overall :tacn:`solve` fails.   

It would be good to review all of the descriptions of evaluation vs application, backtracking, etc. to get the language consistent and simple (refactoring so the common elements are described in one place).  That's a fair bit of work in itself, more than I wanted to do in this PR.

Reply:
I would say "Any :n:`@ltac_expr`\s which do not evaluate to tactic values are skipped." rather than "The :n:`@ltac_expr`\s must evaluate to tactic values.", as `Goal True. solve [ () | 1 | constructor ].` works fine.  It's not clear what "must" means otherwise.





> It would be good to review all of the descriptions of evaluation vs application, backtracking, etc. to get the language consistent and simple (refactoring so the common elements are described in one place).

I agree

> That's a fair bit of work in itself, more than I wanted to do in this PR.

Perhaps open an issue about it, and/or mark it as `todo`, and save it for another PR / another day?  This PR is already a great improvement.

Reply:
> I would say "Any :n:`@ltac_expr`\s which do not evaluate to tactic values are skipped."

OK.  Is that a useful behavior?  Do other compound tactics do that, too?

> Perhaps open an issue about it, and/or mark it as `todo`, and save it for another PR / another day? This PR is already a great improvement.

I'll go for a todo.

Reply:
@JasonGross says:

> OK. Is that a useful behavior? Do other compound tactics do that, too?

On the contrary, I'd consider this an unfortunate design, as it makes it
harder to catch bugs and typos. It's worth mentioning as a "gotcha", I
think. It's bitten me a number of times. (Though I've been bitten even
more often by the fact that, e.g., passing a non-existent lemma to
`rewrite` is non-fatal, and so `first` and `solve` will just skip over any
such branches.)

Other compound tactics have similar behavior, I believe, in that "this is
not a tactic that can be applied to the goal and is a different kind of
value" is just a normal error that induces backtracking like any other
error.

The only place I've found it useful to mix values and tactics is in match,
for example:
```coq
lazymatch x with
| ?v :: _ => v
| _ => fail 0 "not a concrete list with a head:" x
end
```

Reply:
> On the contrary, I'd consider this an unfortunate design, as it makes it
harder to catch bugs and typos. It's worth mentioning as a "gotcha", I
think. It's bitten me a number of times.

I'll make it more of a warning.

Is it worth changing for Ltac1?  Worth a PR?  Would there be compatibility issues?  How about in Ltac2?  I suppose there could a "strict" flag to reject such values.

Does anyone keep a list of Ltac1 weaknesses we'd like to see fixed, though perhaps only in Ltac2?

Reply:
> Is it worth changing for Ltac1? Worth a PR?

It's probably worth a PR, to see the impact and see if there are any
legitimate use cases.  I believe the PR would consist of making "Tactic
expected" an exception that does not use `user_err`, and marking it as
critical / uncatchable by the tactic monad.

> Would there be compatibility issues?

Quite possibly.  My hope is that all the compatibility issues exposed by
such a PR would be genuine bugs in tactic code and this fixable by deleting
the offending line of code, but it may not be the case.  (I can certainly
write some horrible code that relies on the current behavior for no good
reason.). On the other hand, such a PR might reveal a legitimate use-case
for this very questionable feature.

> How about in Ltac2?

I can't recall if it's a warning or an error in Ltac2, but I believe it's
emitted statically at the time the tactic script is type-checked (it's
analogous to writing `"foo"; ()` in OCaml).  This is much more sane.

> I suppose there could a "strict" flag to reject such values.

Indeed, if we go the route of the PR above, such a flag might be useful for
compatibility.

> Does anyone keep a list of Ltac1 weaknesses we'd like to see fixed,
though perhaps only in Ltac2?

@ppedrot might have one somewhere.  I believe most of the insane semantics
issues are already fixed in Ltac2 (at the cost of incompatibly of how one
writes tactic automation)